File: sdks/rust/tests/smoke.rs
use nivrit_sdk::{
    b64_encode, decrypt_secret_value, encapsulate_project_key, encrypt_secret_value,
    generate_user_keypair, NivritSession,
};
use rand::Rng;
use serde_json::json;
use std::time::{SystemTime, UNIX_EPOCH};

fn api_url() -> String {
    std::env::var("NIVRIT_API_URL").unwrap_or_else(|_| "http://localhost:4000".into())
}

async fn api_request(
    method: reqwest::Method,
    path: &str,
    body: Option<serde_json::Value>,
    token: Option<&str>,
) -> serde_json::Value {
    let client = reqwest::Client::new();
    let mut req = client
        .request(method, format!("{}{}", api_url(), path))
        .header("Content-Type", "application/json");
    if let Some(t) = token {
        req = req.header("Authorization", format!("Bearer {}", t));
    }
    if let Some(b) = body {
        req = req.json(&b);
    }
    let resp = req.send().await.unwrap();
    let status = resp.status();
    let text = resp.text().await.unwrap();
    if !status.is_success() {
        panic!("API error {}: {}", status, text);
    }
    serde_json::from_str(&text).unwrap_or(serde_json::Value::Null)
}

#[tokio::test]
async fn smoke_test() {
    let ts = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .as_millis();
    let email = format!("sdk-rust-{}@example.com", ts);
    let password = "Correct-Horse-Battery-Staple!";

    let (public_key, encrypted_private_key, private_key_nonce) =
        generate_user_keypair(password).unwrap();

    let reg = api_request(
        reqwest::Method::POST,
        "/auth/register",
        Some(json!({
            "email": email,
            "password": password,
            "name": "Rust SDK Test",
            "public_key": public_key,
            "encrypted_private_key": encrypted_private_key,
            "private_key_nonce": private_key_nonce,
            "private_key_algorithm": "aes256gcm-v1",
        })),
        None,
    )
    .await;
    println!("registered {}", reg["user"]["email"]);

    let jwt = reg["token"].as_str().unwrap();
    let pat = api_request(
        reqwest::Method::POST,
        "/auth/pat",
        Some(json!({ "name": "rust-sdk-test" })),
        Some(jwt),
    )
    .await;
    println!("created PAT");

    let mut session = NivritSession::from_pat(api_url(), pat["token"].as_str().unwrap(), password)
        .await
        .unwrap();
    println!("session user {}", session.user["email"]);

    let slug = format!("rust-sdk-org-{}", ts);
    let org = session
        .client
        .create_org("Rust SDK Org", &slug)
        .await
        .unwrap();
    println!("created org {}", org["name"]);

    let mut project_key = [0u8; 32];
    rand::rng().fill_bytes(&mut project_key);
    let project_key_b64 = b64_encode(&project_key);

    let encapsulated = encapsulate_project_key(
        &project_key_b64,
        session.user["public_key"].as_str().unwrap(),
    )
    .unwrap();
    let encrypted_project_key = b64_encode(&serde_json::to_vec(&encapsulated).unwrap());

    let mut nonce = [0u8; 12];
    rand::rng().fill_bytes(&mut nonce);
    let project = session
        .client
        .create_project(json!({
            "org_id": org["id"],
            "name": "Rust SDK Project",
            "slug": format!("rust-sdk-project-{}", ts),
            "encrypted_project_key": encrypted_project_key,
            "project_key_nonce": b64_encode(&nonce),
            "project_key_algorithm": "hybrid_x25519_ml_kem_768_aes256gcm_v1",
        }))
        .await
        .unwrap();
    println!("created project {}", project["name"]);

    let env = session
        .client
        .create_environment(project["id"].as_str().unwrap(), "Dev", "dev")
        .await
        .unwrap();
    println!("created environment {}", env["name"]);

    let (ct, n) = encrypt_secret_value("hello-rust-sdk", &project_key_b64).unwrap();
    session
        .client
        .create_secret(
            project["id"].as_str().unwrap(),
            json!({
                "environment_id": env["id"],
                "key": "GREETING",
                "encrypted_value": ct,
                "nonce": n,
                "algorithm": "aes256gcm-v1",
            }),
        )
        .await
        .unwrap();
    println!("created secret");

    let secrets = session
        .list_secrets(project["id"].as_str().unwrap(), env["id"].as_str().unwrap())
        .await
        .unwrap();
    assert_eq!(secrets.len(), 1);
    assert_eq!(secrets[0]["value"], "hello-rust-sdk");
    println!("decrypted secret: {}", secrets[0]["value"]);

    // Also test direct decrypt function
    let direct = decrypt_secret_value(
        secrets[0]["encrypted_value"].as_str().unwrap(),
        secrets[0]["nonce"].as_str().unwrap(),
        &project_key_b64,
    )
    .unwrap();
    assert_eq!(direct, "hello-rust-sdk");

    api_request(
        reqwest::Method::DELETE,
        &format!("/auth/pats/{}", pat["id"].as_str().unwrap()),
        None,
        Some(jwt),
    )
    .await;
    println!("Rust SDK smoke test passed");
}

File: sdks/node/test/smoke.js
const { HelperCrypto, NivritClient, NivritSession } = require('../src');
const crypto = require('crypto');

const API_URL = process.env.NIVRIT_API_URL || 'http://localhost:4000';
const EMAIL = `sdk-node-${Date.now()}@example.com`;
const EMAIL_B = `sdk-node-b-${Date.now()}@example.com`;
const PASSWORD = 'Correct-Horse-Battery-Staple!';

async function request(method, path, body, token) {
  const headers = { 'Content-Type': 'application/json' };
  if (token) headers.Authorization = `Bearer ${token}`;
  const res = await fetch(`${API_URL}${path}`, {
    method,
    headers,
    body: body ? JSON.stringify(body) : undefined,
  });
  const text = await res.text();
  if (!res.ok) throw new Error(`${method} ${path} -> ${res.status}: ${text}`);
  return text ? JSON.parse(text) : undefined;
}

async function main() {
  const helper = new HelperCrypto();
  const material = helper.generateRegistrationMaterial(PASSWORD, EMAIL);

  const reg = await request('POST', '/auth/register', {
    email: EMAIL,
    auth_hash: material.auth_hash,
    name: 'Node SDK Test',
    public_key: material.public_key,
    encrypted_private_key: material.encrypted_private_key,
    private_key_nonce: material.private_key_nonce,
    private_key_algorithm: material.private_key_algorithm,
    recovery_auth_hash: material.recovery_auth_hash,
    encrypted_private_key_recovery: material.encrypted_private_key_recovery,
    private_key_recovery_nonce: material.private_key_recovery_nonce,
    private_key_recovery_algorithm: material.private_key_recovery_algorithm,
  });
  console.log('registered', reg.user.email);

  const pat = await request('POST', '/auth/pat', { name: 'node-sdk-test' }, reg.token);
  console.log('created PAT');

  const session = await NivritSession.fromPat(API_URL, pat.token, PASSWORD, helper);
  console.log('session user', session.user.email);

  const org = await session.client.createOrg({ name: 'Node SDK Org', slug: `node-sdk-org-${Date.now()}` });
  console.log('created org', org.name);

  const projectKey = crypto.randomBytes(32).toString('base64');
  const encapsulated = helper.encapsulateProjectKey(projectKey, session.user.public_key);
  const encryptedProjectKey = Buffer.from(JSON.stringify(encapsulated)).toString('base64');
  const project = await session.client.createProject({
    org_id: org.id,
    name: 'Node SDK Project',
    slug: `node-sdk-project-${Date.now()}`,
    encrypted_project_key: encryptedProjectKey,
    project_key_nonce: crypto.randomBytes(12).toString('base64'),
    project_key_algorithm: 'hybrid_x25519_ml_kem_768_aes256gcm_v1',
  });
  console.log('created project', project.name);

  const env = await session.client.createEnvironment(project.id, { name: 'Dev', slug: 'dev' });
  console.log('created environment', env.name);

  const encrypted = helper.encryptValue('hello-node-sdk', projectKey);
  await session.client.createSecret(project.id, {
    environment_id: env.id,
    key: 'GREETING',
    encrypted_value: encrypted.ciphertext,
    nonce: encrypted.nonce,
    algorithm: 'aes256gcm-v1',
  });
  console.log('created secret');

  const secrets = await session.listSecrets(project.id, env.id);
  if (secrets.length !== 1 || secrets[0].value !== 'hello-node-sdk') {
    throw new Error(`unexpected secrets: ${JSON.stringify(secrets)}`);
  }
  console.log('decrypted secret:', secrets[0].value);

  // --- versioned project-key rotation (ADR 0008) ---------------------------
  const materialB = helper.generateRegistrationMaterial(PASSWORD, EMAIL_B);
  const regB = await request('POST', '/auth/register', {
    email: EMAIL_B,
    auth_hash: materialB.auth_hash,
    name: 'Node SDK Test B',
    public_key: materialB.public_key,
    encrypted_private_key: materialB.encrypted_private_key,
    private_key_nonce: materialB.private_key_nonce,
    private_key_algorithm: materialB.private_key_algorithm,
    recovery_auth_hash: materialB.recovery_auth_hash,
    encrypted_private_key_recovery: materialB.encrypted_private_key_recovery,
    private_key_recovery_nonce: materialB.private_key_recovery_nonce,
    private_key_recovery_algorithm: materialB.private_key_recovery_algorithm,
  });
  console.log('registered second user', regB.user.email);

  // Invited before the rotation: starts out holding only version 1.
  await session.inviteMember(project.id, EMAIL_B, 'member');
  console.log('invited second user to project');

  const patB = await request('POST', '/auth/pat', { name: 'node-sdk-test-b' }, regB.token);
  const sessionB = await NivritSession.fromPat(API_URL, patB.token, PASSWORD, helper);
  const preRotation = await sessionB.getSecret(project.id, env.id, 'GREETING');
  if (preRotation.value !== 'hello-node-sdk') {
    throw new Error(`second user could not decrypt pre-rotation secret: ${JSON.stringify(preRotation)}`);
  }
  console.log('second user decrypted pre-rotation secret');

  const rotated = await session.rotateProjectKey(project.id);
  if (rotated.version !== 2 || rotated.grantedTo !== 2) {
    throw new Error(`unexpected rotation result: ${JSON.stringify(rotated)}`);
  }
  console.log(`rotated project key to version ${rotated.version} (granted to ${rotated.grantedTo} members)`);

  const currentVersion = await session.currentProjectKeyVersion(project.id);
  const currentKey = await session.getProjectKey(project.id);
  const encryptedPostRotation = helper.encryptValue('hello-after-rotation', currentKey);
  await session.client.createSecret(project.id, {
    environment_id: env.id,
    key: 'POST_ROTATION',
    encrypted_value: encryptedPostRotation.ciphertext,
    nonce: encryptedPostRotation.nonce,
    algorithm: 'aes256gcm-v1',
    project_key_version: currentVersion,
  });
  console.log('created post-rotation secret under version', currentVersion);

  // Second user was a current member at rotation time, so they automatically
  // received the new version -- confirm without a fresh login, proving
  // rotateProjectKey's cache update on the rotating side and the grant on
  // the receiving side both worked.
  await sessionB.loadProjectKeys(project.id);
  const postRotationForB = await sessionB.getSecret(project.id, env.id, 'POST_ROTATION');
  if (postRotationForB.value !== 'hello-after-rotation') {
    throw new Error(`second user could not decrypt post-rotation secret: ${JSON.stringify(postRotationForB)}`);
  }
  const preRotationAgainForB = await sessionB.getSecret(project.id, env.id, 'GREETING');
  if (preRotationAgainForB.value !== 'hello-node-sdk') {
    throw new Error(`second user lost access to pre-rotation secret after rotation: ${JSON.stringify(preRotationAgainForB)}`);
  }
  console.log('second user decrypted both pre- and post-rotation secrets after rotation');

  await request('DELETE', `/auth/pats/${patB.id}`, undefined, regB.token);
  await request('DELETE', `/auth/pats/${pat.id}`, undefined, reg.token);
  console.log('revoked PATs');
  console.log('Node SDK smoke test passed.');
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
